### PR TITLE
Attempt at sparse ham acceleration.

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
VQE problems with non-minimal basis sets quickly have Hamiltonians with a lot of terms. An efficient way to evaluate expectation values is representing the Hamiltonian as a sparse matrix, but this transformation is itself slow. One of the most costly operations is summing all the sparse matrices together.

**Description of the Change:**
Analyze the sparsity structure of the terms and sum terms with the same structure in a more efficient way (accessing data directly). 

**Benefits:**
Faster sparse Hamiltonian generation.

**Possible Drawbacks:**
Sparse Hamiltonian generation could be slower in edge cases (e.g. small Hamiltonians with few wires).

**Related GitHub Issues:**

### CODE REVIEW

**Comment by vincentmr** [comment]
Some timing data for an H2 molecule with 

```python
basis = "6-31G"
geometry = geom_struct["H2"](0.74)
symbols = ["H", "H"]
charge = 0
# [...]
%timeit sparse_ham = get_sparse_data(ham)
242 ms ± 13.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) # before
213 ms ± 2.23 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) # after
```

first 1600 terms ( of ~3000 to prevent memory overflow on laptop) 

```python
basis = "CC-PVDZ"
geometry = geom_struct["H2"](0.74)
symbols = ["H", "H"]
charge = 0
# [...]
%timeit sparse_ham = get_sparse_data(ham)
Kernel crashed after a while > 10min # before
7min 51s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each) # before with adaptive buffer
1min 49s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each) # after
```

File: pennylane/ops/qubit/hamiltonian.py [hunk 1]
```diff
@@ -764,3 +781,125 @@ def map_wires(self, wire_map: dict):
...
+    elif ind == 2:
+        data = np.array([-1.0j, 1.0j], dtype=np.complex128)
+        indices = np.array([1, 0], dtype=np.int64)
+    elif ind == 3:
+        data = np.array([1.0, -1.0], dtype=np.complex128)
+        indices = np.array([0, 1], dtype=np.int64)
+    return data, indices
+
+
+def pauli_word_2_mat(vec, coeff=1.0):
```
Comment (albi3ro) [inline pennylane/ops/qubit/hamiltonian.py:876]: Think we could add a version of this to [pauli_airthmetic.py](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/pauli/pauli_arithmetic.py)?

We have `PauliWord.to_mat` which will construct either the dense or sparse matrices.  We could use this logic in that case.

Comment (vincentmr) [reply pennylane/ops/qubit/hamiltonian.py:876]: Yes, I was expecting someone to propose this. I found `pauli_word_to_matrix` in `utils` by it didn't seem used anywhere. I'll discuss with @BorjaRequena and see how to move it there.

Comment (BorjaRequena) [reply pennylane/ops/qubit/hamiltonian.py:876]: I have done a proposal to move this to the `PauliWord` class in #4272. Let me know what you think 😄

**Pull request closed without merging**